src/front/wgsl/mod.rs

```diff
@@ -1520,8 +1520,67 @@ impl Parser {
                     class: crate::ImageClass::Storage(format),
                 }
             }
-            _ => return Err(Error::UnknownType(word)),
-        })
+            name => {
+                if let Some(&handle) = self.lookup_type.get(name) {
+                    let ty = &type_arena[handle];
+                    if ty.name.as_deref() != Some(name) {
+                        return Err(Error::UnknownType(word));
+                    }
+
+                    // This should work for both user-defined structs and type redefinitions.
+                    let inner = match &ty.inner {
```

Comment:
Sorry, I should have clarified. We shouldn't clone here at all.
That is somewhat the reason I don't want to see `derive(Clone)` on the type inner - so that there is a barrier to write the logic like this. See my other comments for how to address this properly.

Reply:
Ah, I see, sorry didn't make that connection. Will fix later this week (currently in midterms).